server/src-lib/Data/Time/Clock/Units.hs

```diff
@@ -25,56 +25,72 @@ You can also go the other way using the constructors rather than the selectors:
 0.5
 @
 
+The 'Read' instances for these types mirror the behavior of the 'RealFrac' instance wrt numeric
+literals for convenient serialization (e.g. when working with env vars):
+
+@
+>>> read "1.2" :: Milliseconds 
+Milliseconds {milliseconds = 0.0012s}
+@
+
 Generally, it doesn’t make sense to pass these wrappers around or put them inside data structures,
-since any function that needs a duration should just accept a 'DiffTime', but they’re useful for
-literals and conversions to/from other types. -}
+since any function that needs a duration should just accept a 'NominalDiffTime', but they’re useful
+for literals and conversions to/from other types. -}
 module Data.Time.Clock.Units
   ( Days(..)
   , Hours(..)
   , Minutes(..)
-  , Seconds
-  , seconds
+  , Seconds(..)
   , Milliseconds(..)
   , Microseconds(..)
   , Nanoseconds(..)
+  -- * Reexports
+  -- | We use 'NominalDiffTime' as the standard type for duration in our code.
```

Comment:
Could you add a comment about why we’ve chosen to use `NominalDiffTime` rather than `DiffTime` here, since the latter is technically more “correct”?

Reply:
I'll add a comment, but I'm not totally sure what to say... it's not clear to me that one is more "correct" than the other as a measure of duration. I suppose `NominalDiffTime` indicates "may have been derived from sketchy math with `UTCTime`s" but that's not what we do

Reply:
`NominalDiffTime` is the correct type for *calendar* duration. It’s the appropriate type when you want to say “this event repeats every day,” since you don’t want it to drift earlier and earlier every time a leap second is introduced! It is not, however, an absolute measure of time.

`DiffTime`, on the other hand, *is* an absolute measure of time. This doesn’t make a significant difference if you’re measuring whole days, or even whole minutes, but if you’re measuring something that takes a second or two, the difference can be meaningful!

Of course, the datatypes don’t innately have any difference; they’re really just numbers. But they’re *supposed* to mean different things, which is why I think it’s surprising to use `NominalDiffTime` to measure wall clock time. I am not strongly opposed to using it this way, but I do think we should have a good justification for doing so.

Could you elaborate slightly on why you prefer using `NominalDiffTime`? It doesn’t seem like it would be enormously difficult to use `DiffTime`, but I assume you have a reason! I just want that reason to end up in the comment.

Reply:
So the way you've described the distinction between `DiffTime` and `NominalDiffTime` is one way to squint at the situation but I don't think it's really reflected in the docs (they are at best ambiguous). In particular I don't think it's very meaningful to declare that `10 :: NominalDiffTime` is not an absolute measure of time. I think they are both _primarily_ durations, and the distinction between the two is only relevant as they relate to `UTCTime` (in the same way that newtype wrappers alter the semantics of a type wrt a particular operation), but again this depends on how you squint.

The comment I have is here https://github.com/hasura/graphql-engine/pull/3622/commits/179af277c5b011faf79cc626902c35db2d94deeb#diff-2ee44e54da5eae501804c04d09d0a270R48

I can't really think of something more interesting to say about this in a comment (but am open to suggestions!). Using `DiffTime` instead just means we have to convert from `DiffTime` to `NominalDiffTime` in various places where we use `addUTCTime` and `diffUTCTime` with no real benefit that I can see.

Reply:
What do you think is ambiguous about the documentation? It’s possible that there’s a reading of the documentation that I’m not seeing, but it seems fairly explicit:

> [`NominalDiffTime`] ignores leap-seconds, so it's not necessarily a fixed amount of clock time. For instance, 23:00 UTC + 2 hours of `NominalDiffTime` = 01:00 UTC (+ 1 day), regardless of whether a leap-second intervened.

1 hour in `NominalDiffTime` is one hour of calendar time, not 3,600 SI seconds (though admittedly the two are usually the same). For *measuring* elapsed time, that isn’t usually the type you want.

Reply:
I think the docs are clumsily referring to the behavior of NominalDiffTime vis a vis UTCTime (which was the motivation for defining a new type as I understand it, it's a sin bin for sketchy time operations). 

In the context of an API in which the only way of constructing a `NominalDiffTime` was `diffUTCTime` (and no comparison operations were supported, and it couldn't be serialized, etc) I think you could say something abstract like "it's not necessarily a fixed amount of clock time" and really mean it.

Reply:
> I think the docs are clumsily referring to the behavior of NominalDiffTime vis a vis UTCTime (which was the motivation for defining a new type as I understand it, it's a sin bin for sketchy time operations).

I agree that this is technically true. There isn’t actually any representational difference between a `DiffTime` and a `NominalDiffTime`, so really the only difference is where the value comes from! It just so happens that you get a `NominalDiffTime` from `diffUTCTime`, which makes it calendar time. I’m in full agreement with you on all of that!

Still, I think the distinction is useful, even if it doesn’t necessarily add much in the way of actual type safety. Really, it’s much the same as writing something like `newtype UserName = UserName Text`. Sure, you *can* store arbitrary `Text` values in a `UserName` value, whether or not they’re actually usernames. But you shouldn’t!

So I think it does make sense to use `DiffTime` here. We can update `withElapsedTime` and `startTimer` to return `DiffTime` values without any additional difficulty, and we’ll get a small bit of type safety in return. If we ever find ourselves needing to deal with intervals of calendar time, we can revisit this—it wouldn’t be hard to generalize these types to be parameterized over the representation type so they work with both `DiffTime` and `NominalDiffTime`.